poetapp/node
# Replace Insight with Bitcoin Core in BlockchainWriter

[PR Process](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-process) - [PR Review Checklist](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-review-checklist)

Fixes https://github.com/poetapp/node/issues/24
Fixes https://github.com/poetapp/node/issues/93
Partly addresses https://github.com/poetapp/node/issues/293
needed-by https://github.com/poetapp/node/issues/22
needed-by https://github.com/poetapp/node/issues/25
needed-by https://github.com/poetapp/node/issues/27

### Release
 Semantic release is enabled for this repository. Make sure you follow the right commit message convention. 
 We're using semantic-release's default — [Angular Commit Message Conventions](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines).

## Description of Changes

This PR replaces the Insight API dependency of the BlockchainWriter module with the Bitcoin Core Wallet, finishing the migration.

### Docker Compose

- A service for regtest was added.
- The testnet service was renamed and its ports were fixed

### Dependencies

`bitcore-lib` and `@types/bitcore-lib` were uninstalled.

### Deleted Tests

Placeholder tests that were introduced to bypass an error in coverage were deleted.

### Rename

This PR started replacing `timestamp` with `anchor`. See https://github.com/poetapp/node/issues/293.

### Configuration

- The default configuration now points to regtest.
- `insightUrl`, `bitcoinAddress` and `bitcoinAddressPrivateKey` were removed.
- No new configuration entries were needed. The onces added in https://github.com/poetapp/node/pull/383 sufficed.

### Fee Estimation

The previous could would always send a fee of `0.001`. This new code is using the [`fundrawtransaction`](https://bitcoincore.org/en/doc/0.16.0/rpc/rawtransactions/fundrawtransaction/) RPC, which estimates fees based on recent blocks.

### How to Test

This is how I tested this change:

- `sudo docker-compose up -d bitcoin-regtest`
- `db.dropDatabase()`
- `npm run build`
- `npm start`
- `npm run test:integration`
- `GET http://localhost:18080/works`
- Verify that response contains a few items.
- After a configured amount of time these items should show `timestamp.ipfsFileHash`, `timestamp.ipfsDirectoryHash` and `transactionId`. These three properties will appear one by one. Keep refreshing until the show.
- `bitcoin-cli generate 1`
- Wait the configured amount of time until BlockchainReader picks up the generated block
- `GET http://localhost:18080/works` again
- The works in the response should now contain a full `timestamp`, with `blockHash`, `blockHeight`, etc.

I ran this test both in regtest and testnet. 

One ideal test I haven't done is node syncing: deploy this code and the dependencies to two different machines and confirm they sync correctly.

## Things Being Left Out

I'm leaving out of this PR a few things since it's actually working as it is and it's blocking several other PRs.

- Update documentation
- https://github.com/poetapp/node/issues/157
- Unit tests
- Possible refactors or code improvements
